Andrew.ardito/span2metrics (#9263)
DataDog/documentation: content/en/tracing/generate_metrics/_index.md

---
title: Generate Metrics from Spans
kind: documentation
description: 'Generate custom metrics from ingested spans.'
aliases:
- /tracing/span_to_metrics/
further_reading:
    - link: 'tracing/trace_retention_and_ingestion'
      tag: "Documentation"
      text: 'Customize trace ingestion and retain important traces.'
    - link: 'tracing/trace_search_and_analytics/query_syntax'
      tag: "Documentation"
      text: 'Use Analytics queries and monitors based on retained traces.'
---
<div class="alert alert-warning">
This feature is currently in private beta. <a href="https://docs.google.com/forms/d/1RlT0FNdFjiEzkQgxiCf77ugpW0w5a17X7JQ7E286jM4">Fill out this form</a> to request this feature for your account or be notified when it becomes Generally Available.
</div>

> ## Generate Span-based Metrics

Author: SGTM!

> Tracing without Limits™ allows you to generate metrics from 100% of ingested spans, regardless of if they are indexed by a [retention filter][1].

Reviewer: ```suggestion
With Tracing without Limits™, you can generate metrics from 100% of ingested spans, regardless of whether they are indexed by a [retention filter][1].
```

> These metrics can be paired with retention filters and Analytics monitors, or used on their own.

Reviewer: ```suggestion
You can pair these metrics with retention filters and Analytics monitors, or use them on their own.
```

| Retention Period                     | 15 months                    | 15 days             |
| Anomaly Detection                           | Create an [Anomaly Monitor][3] based on generated metrics.                            | Use Analytics to compare behavior over the past 15 days, and drill into complete traces to investigate root cause.                         |
| Investigation of matching traces with full context                          | N/A - Custom Metrics will not result in any retention of associated traces.                            | Keep exactly the traces relevant to your business context with [retention filters][1].                            |
| Granularity of behavior                           | Create custom metrics for important endpoints or other low-cardinality groups.                        | Use [Search and Analytics][4] for specific endpoints, or use the 'Group By' option in [Analytics][5].                    |
| Forecasting or complex mathematics                          | Create a [Forecast monitor][6] based on generated metrics.                          |   N/A                            |

> To generate metrics from span, navigate from the [APM Setup and Configuration][7] page to [Generate Metrics][8] on the top of the page, and click the “+ New Metric” button.

Reviewer: ```suggestion
To generate metrics from spans, on the [APM Setup and Configuration][7] page select the [Generate Metrics][8] tab, and click the **New Metric** button.
```

<br>

{{< img src="tracing/span_to_metrics/GenerateMetrics.png" style="width:100%;" alt="Generate metrics from ingested spans" >}}

> ### Creating a New Span-based Metric

Reviewer: ```suggestion
### Creating a span-based metric
```
(If you create something, it's new, and you don't have to say so)

{{< img src="tracing/span_to_metrics/createspantometrics.png" style="width:100%;" alt="How to create a metric" >}}

> 1. Define the metric query: Start by adding a query for filtering to your required dataset. The [query syntax][9] is the same as APM Search and Analytics.

Reviewer: ```suggestion
1. **Define the metric query:** Start by adding a query for filtering to your required dataset. The [query syntax][9] is the same as APM Search and Analytics.
```

> 1. Define the field you want to track: Select * to generate a count of all spans matching your query or enter an attribute (e.g., `@cassandra_row_count`) to aggregate a numeric value and create its corresponding count, min, max, sum, and avg aggregated metrics. If the attribute type is a measure, the value of the metric is the value of the span attribute.

Reviewer: ```suggestion
1. **Define the field you want to track:** Select `*` to generate a count of all spans matching your query or enter an attribute (for example, `@cassandra_row_count`) to aggregate a numeric value and create its corresponding count, minimum, maximum, sum, and average aggregated metrics. If the attribute type is a measure, the value of the metric is the value of the span attribute.
```

> 1. Specify the group-by dimension: By default, metrics generated from spans will not have any tags unless explicitly added. Any attribute or tag that exists in your spans can be used to create metric Tags.

Reviewer: ```suggestion
1. **Specify the group-by dimension:** By default, metrics generated from spans will not have any tags unless explicitly added. Any attribute or tag that exists in your spans can be used to create metric tags.
```

> 1. Check the Live Analytics and Search Query preview: You can view the impact of your query in real-time on the data visualization, and the matching spans considered for your query in a live preview.

Reviewer: ```suggestion
1. **Check the Live Analytics and Search Query preview:** You can view the impact of your query in real-time on the data visualization, and the matching spans considered for your query in a live preview.
```

> 1. Name your metric: Metric names must follow the [metric naming convention][10]. Metric names beginning with `trace.*` are not permitted and will not be saved.

Reviewer: ```suggestion
1. **Name your metric:** Metric names must follow the [metric naming convention][10]. Metric names that start with `trace.*` are not permitted and will not be saved.
```
Do you mean "that start with `trace.` maybe?

Author: Yeah I think it's okay to have it as `trace.*` since we refer to them as that in a lot of places and it's the wildcard.  But I like the reword!

**Important Note**: Span-based metrics are considered custom metrics and billed accordingly. Avoid grouping by unbounded or extremely high cardinality attributes like timestamps, user IDs, request IDs, or session IDs to avoid impacting your billing.

> ### Updating Existing Span-based Metrics

Reviewer: ```suggestion
### Updating existing span-based metrics
```

{{< img src="tracing/span_to_metrics/editspantometrics.png" style="width:100%;" alt="Edit an existing metrics" >}}

After a metric is created, only two fields can be updated:

| Field                                 | Reason                                                                                                             |
|----------------------------------------|-------------------------------------------------------------------------------------------------------------------------|
| Stream filter query                  | Change the set of matching spans to be aggregated into metrics.            |

> | Aggregation groups             | Update the tags of manage the cardinality of generated metrics.                                                     |

Reviewer: ```suggestion
| Aggregation groups             | Update the tags to manage the cardinality of generated metrics.                                                     |
```

> **Note**: To change the metric type or name, a new metric must be created

Reviewer: ```suggestion
**Note**: To change the metric type or name, create a new metric and delete the old one.
```
